### PR TITLE
feat(new_split_chunks): support `max_size`

### PR DESCRIPTION
**Pull request by hyf0**
## Summary

The last important piece of `SplitChunksPlugin`. I would focus on migrating tests after this PR is merged.

### <samp>🤖 Generated by Copilot at f6ba0f0</samp>

This pull request adds new features and improvements to the split chunks plugin for `rspack`, a Rust-based web packer. It allows setting maximum sizes and a fallback cache group for the chunks, and refactors the options conversion and the plugin logic. It also updates the TypeScript configuration schema and types, and adds a new test case for the maximum size option.

<details open=true>
  <summary><h2>Walkthrough</h2></summary>

### <samp>🤖 Generated by Copilot at f6ba0f0</samp>

*  Add new options for the split chunks plugin to control the maximum sizes for chunks based on their type and name, and the fallback cache group for unmatched chunks ([link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-0ffdb52a1a2c06dd32d3378adb43a463da9c779f285d0b41e8bad0fcc6582968R12),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-0ffdb52a1a2c06dd32d3378adb43a463da9c779f285d0b41e8bad0fcc6582968L27-R30),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-0ffdb52a1a2c06dd32d3378adb43a463da9c779f285d0b41e8bad0fcc6582968L113-R116),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-0ffdb52a1a2c06dd32d3378adb43a463da9c779f285d0b41e8bad0fcc6582968L126-R132),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-0ffdb52a1a2c06dd32d3378adb43a463da9c779f285d0b41e8bad0fcc6582968L151-R154),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-0ffdb52a1a2c06dd32d3378adb43a463da9c779f285d0b41e8bad0fcc6582968L161-R187),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-0ffdb52a1a2c06dd32d3378adb43a463da9c779f285d0b41e8bad0fcc6582968L184-R209),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-0ffdb52a1a2c06dd32d3378adb43a463da9c779f285d0b41e8bad0fcc6582968L198-R270),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-bb77f4a776a00a1377e31ab5f38a079c93e6767ba399c09227c2def2a2067056R21),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-0aab5b8bd0e625e143e1d9cfe68dbf4e7339d9646deffa103a03eaef98c527ebR34-R35),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-5333b4dda3fcc6f86cfcd38f67764a1123d5d0a159da62f881f7624d686b45a9L8-R11),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-5333b4dda3fcc6f86cfcd38f67764a1123d5d0a159da62f881f7624d686b45a9L58-R60),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-5333b4dda3fcc6f86cfcd38f67764a1123d5d0a159da62f881f7624d686b45a9R75-R132),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-5333b4dda3fcc6f86cfcd38f67764a1123d5d0a159da62f881f7624d686b45a9R160-R169),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-68521d603ace0ec833298b9dd6d5a2afd08219d5003d72f1022193ae9a820812L15-R15),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-d37e7d6b15dacdb49c0f24c23ab58b699e5e218e9171a910bb3b4637164e4a5aR1-R322),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-8ae45ec9606865dac58ced163f0e1e7efb9b21d0d014132b5a2de6587f1d4a4aR3),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-8ae45ec9606865dac58ced163f0e1e7efb9b21d0d014132b5a2de6587f1d4a4aL8-R34),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-8ae45ec9606865dac58ced163f0e1e7efb9b21d0d014132b5a2de6587f1d4a4aL32-R57),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-8ae45ec9606865dac58ced163f0e1e7efb9b21d0d014132b5a2de6587f1d4a4aL86-R113),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-8ae45ec9606865dac58ced163f0e1e7efb9b21d0d014132b5a2de6587f1d4a4aL106-R132),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-8ae45ec9606865dac58ced163f0e1e7efb9b21d0d014132b5a2de6587f1d4a4aR158-R164),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-5f5bb1aeb2a977ff30c5ab822c79f09b76476c503462939c69beb9e4d7552123L497-R500),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-5f5bb1aeb2a977ff30c5ab822c79f09b76476c503462939c69beb9e4d7552123L509-R516),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-d7e2c50a4845569f35f387513dc5a6e542e9d5dde107294fd55c5ff4e8a72c88R980-R985),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-d7e2c50a4845569f35f387513dc5a6e542e9d5dde107294fd55c5ff4e8a72c88R1009-R1028),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-d7e2c50a4845569f35f387513dc5a6e542e9d5dde107294fd55c5ff4e8a72c88R1103-R1111),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-790d97d1c7c2bfbbe7720a38355ec1e0d462206a5ddb9826c6aa8bff7090ce0fR16),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-790d97d1c7c2bfbbe7720a38355ec1e0d462206a5ddb9826c6aa8bff7090ce0fR597-R600),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-790d97d1c7c2bfbbe7720a38355ec1e0d462206a5ddb9826c6aa8bff7090ce0fR610-R612))
* Implement the logic for splitting the chunks that are bigger than the maximum size limit, using a deterministic grouping algorithm and a `ChunkToBeSplit` struct ([link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-d37e7d6b15dacdb49c0f24c23ab58b699e5e218e9171a910bb3b4637164e4a5aR1-R322),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-8ae45ec9606865dac58ced163f0e1e7efb9b21d0d014132b5a2de6587f1d4a4aL106-R132))
* Add a new test case `max-size-casing` to verify the maximum size option and the case-insensitive chunk names, with four modules and one dynamic chunk ([link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-cbc73055280758761f5065b408e6c9a2095adac809ce7f94b0c7330f38a19c10R1-R9),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-38661179bc690ae39740a2a72751635a028f6ca43041205798a25b9b5b24ad7aR1),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-d1bb817798cf50c069220abd515bb79f9358127749e679fb8a88db5646422f82R1),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-90d0e9609e970c43d37c38f841a4c75cc0b16e96bad5c6a04afaef30c960ebb4R1),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-a8c17f418bc1e1815ee4a46e07d1890521b88d71406513d9140b7f17d1371deaR1),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-2d9cbaa7193b17c20b16b95f55ccd57ed4626ed715f9d4c9b1c70f374f24ef81R1-R10))
* Add tracing statements to log the reasons for ignoring or removing module groups, and the prepared module group map ([link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-491710d760a9177c165f505436fcb4f19f6f00caaeb2cd888f6377d881827d23R72-R77),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-491710d760a9177c165f505436fcb4f19f6f00caaeb2cd888f6377d881827d23R90-R96),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-aea417a2412d75c4062fa5334fcf6a8598277f96282261dc9769e39f63be059dR20-R24),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-aea417a2412d75c4062fa5334fcf6a8598277f96282261dc9769e39f63be059dR40-R47),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-aea417a2412d75c4062fa5334fcf6a8598277f96282261dc9769e39f63be059dR90-R93),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-8ae45ec9606865dac58ced163f0e1e7efb9b21d0d014132b5a2de6587f1d4a4aL32-R57))
* Remove unnecessary or redundant tracing statements and attributes ([link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-9fe0bd2bb40e0e56daf84e7c76107d80d06b5b306b639db5194b24e5daaca8c7L20),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-df5e22b8430b27c0d6936f0c0a74a3f449f5b84a3b60d7c84eacb5fb719e88b9L57),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-aea417a2412d75c4062fa5334fcf6a8598277f96282261dc9769e39f63be059dL69))
* Add a `Clone` trait derivation and a `merge` method to the `SplitChunkSizes` struct, and make the inner field `pub(crate)` ([link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-5333b4dda3fcc6f86cfcd38f67764a1123d5d0a159da62f881f7624d686b45a9L58-R60),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-5333b4dda3fcc6f86cfcd38f67764a1123d5d0a159da62f881f7624d686b45a9R75-R132))
* Add a `hidePathInfo` option to the split chunks plugin to control whether to hide the path info in the chunk names, and copy it from the TypeScript options to the Rust options ([link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-d7e2c50a4845569f35f387513dc5a6e542e9d5dde107294fd55c5ff4e8a72c88R980-R985),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-5f5bb1aeb2a977ff30c5ab822c79f09b76476c503462939c69beb9e4d7552123L509-R516))
* Use a new function `create_chunk_filter_from_str` to create a `ChunkFilter` from a string option, instead of matching on the string value ([link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-0ffdb52a1a2c06dd32d3378adb43a463da9c779f285d0b41e8bad0fcc6582968L126-R132),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-0ffdb52a1a2c06dd32d3378adb43a463da9c779f285d0b41e8bad0fcc6582968L184-R209))
* Use a new function `create_sizes` to create a `SplitChunkSizes` from an optional `f64` value, using the `default_size_types` array ([link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-0ffdb52a1a2c06dd32d3378adb43a463da9c779f285d0b41e8bad0fcc6582968L151-R154),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-0ffdb52a1a2c06dd32d3378adb43a463da9c779f285d0b41e8bad0fcc6582968L161-R187),[link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-0ffdb52a1a2c06dd32d3378adb43a463da9c779f285d0b41e8bad0fcc6582968L198-R270))
* Use the `derivative` crate to derive traits with custom options, and the `FxHashSet` type from the `rustc_hash` crate for fast hash sets ([link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-5333b4dda3fcc6f86cfcd38f67764a1123d5d0a159da62f881f7624d686b45a9L8-R11))
* Use the `rspack_util` crate as a dependency for the `rspack_plugin_split_chunks_new` crate, which provides some utility functions and types ([link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-bb77f4a776a00a1377e31ab5f38a079c93e6767ba399c09227c2def2a2067056R21))
* Use the `rspack_plugin_split_chunks_new` module prefix instead of the `new_split_chunks_plugin` alias ([link](https://github.com/web-infra-dev/rspack/pull/3171/files?diff=unified&w=0#diff-0ffdb52a1a2c06dd32d3378adb43a463da9c779f285d0b41e8bad0fcc6582968L184-R209))

</details>

### CODE REVIEW

File: crates/rspack_plugin_split_chunks_new/src/plugin/max_size.rs [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+    let infos = chunks_to_be_split
+      .into_par_iter()
+      .filter_map(|info| {
+        let ChunkToBeSplit {
+          chunk,
+          allow_max_size,
+          min_size,
+        } = &info;
+        let results = deterministic_grouping_for_modules(
+          &compilation.lock().expect("Should not panic"),
```
Comment (h-a-n-a) [inline crates/rspack_plugin_split_chunks_new/src/plugin/max_size.rs:255]: Will this create additional overhead comparing to plain iterator?

Comment (hyf0) [reply crates/rspack_plugin_split_chunks_new/src/plugin/max_size.rs:255]: It's intensional. I want this code to be parallel-safe so I could refactor them to run in parallel later.

**Review by h-a-n-a — APPROVED** [review]
